### PR TITLE
Add run_policy to ElasticTask

### PR DESCRIPTION
**Pull request by bgedik**
## Why are the changes needed?

`Elastic` tasks currently do not expose a run policy. As such, the run policy of a `PyTorchJob` created via an `Elastic` task cannot be configured. The same functionality exists for a `PyTorchJob` created via a `PyTorch` task. 

## What changes were proposed in this pull request?

We are adding the same run policy configuration support from `PyTorch` tasks to `Elastic` tasks.

## How was this patch tested?

Unit tests added

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [X] All new and existing tests passed.
- [X] All commits are signed-off.

### CODE REVIEW

**Comment by bgedik** [comment]
cc @fg91

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 1]
```diff
@@ -137,6 +138,7 @@ class Elastic(object):
     monitor_interval: int = 5
     max_restarts: int = 0
     rdzv_configs: Dict[str, Any] = field(default_factory=dict)
+    run_policy: Optional[RunPolicy] = field(default_factory=lambda: None)
```
Comment (fg91) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:142]: For the `rdzv_configs` which defaults to an empty dict, using a default factory is required to ensure that each instance gets a unique copy of the mutable default value.

I don't think anything speaks against using `run_policy: Optional[RunPolicy] = None` though, or am I overlooking something here?

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py [hunk 2]
```diff
@@ -427,6 +429,14 @@ def execute(self, **kwargs) -> Any:
 
         return exception_scopes.user_entry_point(self._execute)(**kwargs)
 
+    def _convert_run_policy(self, run_policy: RunPolicy) -> kubeflow_common.RunPolicy:
```
Comment (fg91) [inline plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:432]: This method is exactly the same as in `PytorchFunctionTask`. It also doesn't need access to `self`. Could you please remove the duplication by removing the method from the classes and instead turn it into a standalone function called e.g. `_convert_run_policy_to_flyte_idl`?

File: plugins/flytekit-kf-pytorch/tests/test_elastic_task.py [hunk 3]
```diff
@@ -187,3 +188,38 @@ def wf(recoverable: bool):
...
+        clean_pod_policy=CleanPodPolicy.ALL,
+        ttl_seconds_after_finished=10 * 60,
+        active_deadline_seconds=36000,
+        backoff_limit=None,
+    )
+
+    pytorch_tasks = []
+    original_init = PytorchElasticFunctionTask.__init__
+
+    def mock_init(self, *args, **kwargs) -> None:
```
Comment (fg91) [inline plugins/flytekit-kf-pytorch/tests/test_elastic_task.py:207]: Thanks for adding a test that ensures that the run policy is correctly set by `.get_custom()`.

I find the mocking logic unnecessarily complex. [This](https://github.com/flyteorg/flytekit/blob/2719b349cd6ac9b18cad858ff2fc6d6f79986b1e/plugins/flytekit-kf-pytorch/tests/test_pytorch_task.py#L143) test for the `PytorchFunctionTask` shows how you can call `.get_custom` on a task without mocking the constructor. Can you please simplify it here?

**Review by fg91 — CHANGES_REQUESTED** [review]
Thank you for adding the `RunPolicy` to the `Elastic` task config. I tested your PR and can confirm that the `RunPolicy` is correctly applied to the `PytorchJob` object in the cluster and that e.g. a "clean all" run policy works as expected.

I have three small change requests, see comments above. Could you please address them and re-request a review? After that I'll approve 🙏

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 2 above)
Comment (bgedik) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:432]: Addressed

File: plugins/flytekit-kf-pytorch/tests/test_elastic_task.py (hunk 3 above)
Comment (bgedik) [reply plugins/flytekit-kf-pytorch/tests/test_elastic_task.py:207]: Addressed

File: plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py (hunk 1 above)
Comment (bgedik) [reply plugins/flytekit-kf-pytorch/flytekitplugins/kfpytorch/task.py:142]: Addressed

**Comment by bgedik** [comment]
> Thank you for adding the `RunPolicy` to the `Elastic` task config. I tested your PR and can confirm that the `RunPolicy` is correctly applied to the `PytorchJob` object in the cluster and that e.g. a "clean all" run policy works as expected.
> 
> I have three small change requests, see comments above. Could you please address them and re-request a review? After that I'll approve 🙏

Hi @fg91 Addressed your comments

**Comment by fg91** [comment]
> Hi @fg91 Addressed your comments

Can you please sign-off the commits/fix DCO?

**Comment by bgedik** [comment]
> > Hi @fg91 Addressed your comments
> 
> Can you please sign-off the commits/fix DCO?

@fg91  Signed

**Pull request closed without merging**